### ISSUE TITLE
[4.0.0] regression - NumberFormatException: For input string: "$"

### ISSUE THREAD

**Issue by francisdb** [jmrozanec/cron-utils#89]
We have a test like this

```scala
  "throw an exception for invalid cron expression with unknown char" in {
      parse("* * * * $ ?") must throwA[IllegalArgumentException](
        Pattern.quote("Invalid chars in expression! Expression: $ Invalid chars: $")
      )
    }
```

but with 4.0.0 this now fails with a IllegalArgumentException with message `java.lang.NumberFormatException: For input string: "$"'`
This is way to technical to show to an enduser

this seems to be a regression from #11 where this was fixed:
https://github.com/jmrozanec/cron-utils/commit/7bb8f323d09a47776ae30a372ab721c32be80708#diff-6effbc1e4c5d8c95168d63f715564b44R149

Most of the tests only test for exceptions thrown without validating messages:
https://github.com/jmrozanec/cron-utils/commit/7bb8f323d09a47776ae30a372ab721c32be80708#diff-b9ebcbf03bf1827c5a96cb483e84ea62R37

**Comment by jmrozanec** [comment]
@francisdb thank you for reporting this. Do you have some more tests similar to the reported one, we could merge into our codebase?

**Comment by francisdb** [comment]
just pasting this blob, we have a wrapper class we use

```scala
 "Validating a cron expression" should {

    "succeed for a valid cron expression 0 0/15 * * * ?" in {
      CronUtil.validateCronExpression("0 0/15 * * * ?", true)
      success
    }

    "throw an exception when given an invalid cron expression for quartz" in {
      CronUtil.validateCronExpression("0 30 17 4 1 * 2016", false) must throwA[IllegalArgumentException](
        Pattern.quote("Both, a day-of-week AND a day-of-month parameter, are not supported.")
      )
    }

    "throw an exception for invalid cron expression with unknown char" in {
      CronUtil.validateCronExpression("* * * * $ ?", false) must throwA[IllegalArgumentException](
        // TODO should be better with https://github.com/jmrozanec/cron-utils/issues/89
        Pattern.quote("""java.lang.NumberFormatException: For input string: "$"""")
      )
    }

    "for an invalid quartz cron expression throw an exception with message containing the actual expression" in {
      skipped("https://github.com/jmrozanec/cron-utils/issues/90")
      CronUtil.validateCronExpression("0/1 * * * * *", false) must throwA[IllegalArgumentException](
        Pattern.quote("""Invalid cron expression: 0/1 * * * * *. Both, a day-of-week AND a day-of-month parameter, are not supported.""")
      )
    }
  }

  "Calculating the minimum interval for a cron expression" should {

    "be 1 for * * * * * ?" in {
      CronUtil.getShortestPeriodSeconds("* * * * * ?", Duration.ZERO) must be equalTo Duration.ofSeconds(1)
    }

    "be 1 for *   *    *  *       * ? (ignore whitespace)" in {
      CronUtil.getShortestPeriodSeconds("* * * * * ?", Duration.ZERO) must be equalTo Duration.ofSeconds(1)
    }

    "be 1 for 0/1 * * * * ?" in {
      CronUtil.getShortestPeriodSeconds("0/1 * * * * ?", Duration.ZERO) must be equalTo Duration.ofSeconds(1)
    }

    "be 1 for */1 * * * * ?" in {
      CronUtil.getShortestPeriodSeconds("*/1 * * * * ?", Duration.ZERO) must be equalTo Duration.ofSeconds(1)
    }

    "be 60 for 0 * * * * ?" in {
      CronUtil.getShortestPeriodSeconds("0 * * * * ?", Duration.ZERO) must be equalTo Duration.ofMinutes(1)
    }

    "be 60 for 0 */1 * * * ?" in {
      CronUtil.getShortestPeriodSeconds("0 */1 * * * ?", Duration.ZERO) must be equalTo Duration.ofMinutes(1)
    }

    "be 300 for 0 */5 * * * ?" in {
      CronUtil.getShortestPeriodSeconds("0 */5 * * * ?", Duration.ZERO) must be equalTo Duration.ofMinutes(5)
    }

    "be 3600 for 0 0 * * * ?" in {
      CronUtil.getShortestPeriodSeconds("0 0 * * * ?", Duration.ZERO) must be equalTo Duration.ofHours(1)
    }

    "be 10800 for 0 0 */3 * * ?" in {
      CronUtil.getShortestPeriodSeconds("0 0 */3 * * ?", Duration.ZERO) must be equalTo Duration.ofHours(3)
    }

    "be 86400 for 0 0 0 * * ?" in {
      CronUtil.getShortestPeriodSeconds("0 0 0 * * ?", Duration.ofDays(1)) must be equalTo Duration.ofDays(1)
    }

    "throw an exception for invalid cron expression * *" in {
      CronUtil.getShortestPeriodSeconds("* *", Duration.ZERO) must throwA[IllegalArgumentException](
        Pattern.quote("Cron expression contains 2 parts but we expect one of [6, 7]")
      )
    }

    "throw an exception for invalid cron expression */ * * * * ?" in {
      CronUtil.getShortestPeriodSeconds("*/ * * * * ?", Duration.ZERO) must throwA[IllegalArgumentException](
        Pattern.quote("Missing steps for expression: */")
      )
    }

    "throw an exception for invalid cron expression with unknown char" in {
      CronUtil.getShortestPeriodSeconds("* * * * $ ?", Duration.ZERO) must throwA[IllegalArgumentException](
        // TODO should be better with https://github.com/jmrozanec/cron-utils/issues/89
        Pattern.quote("""java.lang.NumberFormatException: For input string: "$"""")
      )
    }
```

**Comment by francisdb** [comment]
But asserting on error messages for the current test would already help a lot

**Comment by jmrozanec** [comment]
@francisdb Great! Thanks!

**Comment by jmrozanec** [comment]
@francisdb Fixed!

**Comment by jmrozanec** [comment]
Fix published in [release 4.0.1](https://github.com/jmrozanec/cron-utils/releases/tag/4.0.1)
